Update docs with `save`, `push_to_hub`, and `evaluator` (#85)
huggingface/evaluate: docs/source/a_quick_tour.mdx

This solution allows 🤗 Evaluate to perform distributed predictions, which is important for evaluation speed in distributed settings. At the same time, you can also use complex non-additive metrics without wasting valuable GPU or CPU memory.

> ## Save and push to hub

Reviewer: ```suggestion
## Save and push to the Hub
```

> Saving and sharing evaluation results is an important step. We provide the [`evaluate.save()`] function to easily save metrics result. You can either point pass a specific filename or a directory. In the latter case the results are saved in a file with an automatically created file name. Besides the directory or file name the function takes any key-value pairs as inputs and stores them in a JSON file.

Reviewer: ```suggestion
Saving and sharing evaluation results is an important step. We provide the [`evaluate.save()`] function to easily save metrics results. You can either pass a specific filename or a directory. In the latter case, the results are saved in a file with an automatically created file name. Besides the directory or file name, the function takes any key-value pairs as inputs and stores them in a JSON file.
```

```py
result = accuracy.compute(references=[0,1,0,1], predictions=[1,0,0,1])

hyperparams = {"model": "bert-base-uncased"}
evaluate.save("./results/"experiment="run 42", **result, **hyperparams)
>>> PosixPath('results/result-2022_05_30-22_09_11.json')
```

The content of the JSON file look like the following:

```json
{
    "experiment": "run 42",
    "accuracy": 0.5,
    "model": "bert-base-uncased",
    "_timestamp": "2022-05-30T22:09:11.959469",
    "_git_commit_hash": "123456789abcdefghijkl",
    "_evaluate_version": "0.1.0",
    "_python_version": "3.9.12 (main, Mar 26 2022, 15:51:15) \n[Clang 13.1.6 (clang-1316.0.21.2)]",
    "_interpreter_path": "/Users/leandro/git/evaluate/env/bin/python"
}
```

> In addition to the specified fields it also contains information about the system useful for reproducing the results. 

Reviewer: ```suggestion
In addition to the specified fields, it also contains useful system information for reproducing the results. 
```

> Besides storing the results locally we also want to report them on the model's repository on the hub. With the [`evaluate.push_to_hub()`] you can easily report an evaluation results to the model's repository:

Reviewer: ```suggestion
Besides storing the results locally, you should report them on the model's repository on the Hub. With the [`evaluate.push_to_hub()`] function, you can easily report evaluation results to the model's repository:
```

```py
evaluate.push_to_hub(
  model_id="huggingface/gpt2-wikitext2",  # model repository on hub
  metric_value=0.5,                       # metric value
  metric_type="bleu",                     # metric name, e.g. accuracy.name
  metric_name="BLEU",                     # pretty name which is displayed
  dataset_type="wikitext",                # dataset name on the hub
  dataset_name="WikiText",                # pretty name
  dataset_split="test",                   # dataset split used
  task_type="text-generation",            # task id, see https://github.com/huggingface/datasets/blob/master/src/datasets/utils/resources/tasks.json
  task_name="Text Generation"             # pretty name for task
)
```

## Evaluator

> The [`evaluate.evaluator()`] provides integrate the inference part of evaluation and only requires a model, dataset, metric in contrast to the metrics in `EvaluationModule`s that require the model's predictions. As such it is easier to evaluate a model on a dataset with a given metric. To make that possible it uses the `Pipeline` abstraction from `transformers`. However, you can use your own framework as long as it follows the `Pipeline` interface.

Reviewer: The first sentence is a bit hard to read

Reviewer: Much better! :fire:

> To make an evaluation with the `evaluator` let's load a pipeline with an model trained on IMDb, the IMDb test split and the accuracy metric. 

Reviewer: Maybe mention this is a transformers pipeline as people might not know about it. I would also mention this works for other libraries

```python
from transformers import pipeline
from datasets import load_dataset
from evaluate import evaluator
import evaluate

pipe = pipeline("text-classification", model="lvwerra/distilbert-imdb", device=0)
data = load_dataset("imdb", split="test").shuffle().select(range(1000))
metric = evaluate.load("accuracy")
```

> Then we can create an evaluator for text classification and pass the three objects to the `compute` method. With the label mapping we provide a method to align the pipeline outputs with the label column in the dataset:

Reviewer: ```suggestion
Then you can create an evaluator for text classification and pass the three objects to the `compute` method. With the label mapping `evaluate` provides a method to align the pipeline outputs with the label column in the dataset:
```

```python
eval = evaluator("text-classification")

results = eval.compute(model_or_pipeline=pipe, data=data, metric=metric, label_mapping={"NEGATIVE": 0, "POSITIVE": 1},)

print(results)
>>> {'accuracy': 0.934}
```

> Calculating the value of the metric alone is often not enough to know if a model performs significantly better than another one. With _bootstrapping_ we compute confidence intervales and the standard error which helps estimate how stable a score is:

Reviewer: ```suggestion
Calculating the value of the metric alone is often not enough to know if a model performs significantly better than another one. With _bootstrapping_ `evaluate` computes confidence intervals and the standard error which helps estimate how stable a score is:
```

### Distributed evaluation

Computing metrics in a distributed environment can be tricky. Metric evaluation is executed in separate Python processes, or nodes, on different subsets of a dataset. Typically, when a metric score is additive (`f(AuB) = f(A) + f(B)`), you can use distributed reduce operations to gather the scores for each subset of the dataset. But when a metric is non-additive (`f(AuB) ≠ f(A) + f(B)`), it's not that simple. For example, you can't take the sum of the [F1](https://huggingface.co/spaces/evaluate-metric/f1) scores of each data subset as your **final metric**.

A common way to overcome this issue is to fallback on single process evaluation. The metrics are evaluated on a single GPU, which becomes inefficient.

> 🤗 Evaluate solves this issue by only computing the final metric on the first node. The predictions and references are computed and provided to the metric separately for each node. These are temporarily stored in an Apache Arrow table, avoiding cluttering the GPU or CPU memory. When you are ready to `compute` the final metric, the first node is able to access the predictions and references stored on all the other nodes. Once it has gathered all the predictions and references, `compute` will perform the final metric evaluation.

Reviewer 2: missing parentheses for `compute()`
